File: rust/src/compare_binary_trees.rs
pub struct BinaryNode<T> {
    pub value: T,
    pub left: Option<Box<BinaryNode<T>>>,
    pub right: Option<Box<BinaryNode<T>>>,
}

pub fn compare<T: PartialEq>(a: Option<&BinaryNode<T>>, b: Option<&BinaryNode<T>>) -> bool {
    match (a, b) {
        (None, None) => true,
        (Some(node_a), Some(node_b)) => {
            node_a.value == node_b.value && compare(node_a.left.as_ref(), node_b.left.as_ref()) && compare(node_a.right.as_ref(), node_b.right.as_ref())
        }
        _ => false,
    }
}

File: java/src/CompareBinaryTrees.java
package src;

public class BinaryTreeComparison {

    public static class BinaryNode<T> {
        T value;
        BinaryNode<T> left;
        BinaryNode<T> right;

        public BinaryNode(T value, BinaryNode<T> left, BinaryNode<T> right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }
    }

    public static boolean compare(BinaryNode<Integer> a, BinaryNode<Integer> b) {
        // Base case 1: Structural check
        if (a == null && b == null) {
            return true;
        }

        // Base case 2: Structural check
        if (a == null || b == null) {
            return false;
        }

        // Base case 3: Value check
        if (!a.value.equals(b.value)) {
            return false;
        }

        return compare(a.left, b.left) && compare(a.right, b.right);
    }
}
